monix-execution/jvm/src/main/scala/monix/execution/workstealing/WorkStealingThreadPool.scala

```diff
@@ -0,0 +1,458 @@
+    *
+    * @param random a reference to an uncontended source of randomness, to be
+    *               passed along to the striped concurrent queues when executing
+    *               their enqueue operations
+    */
+  private[workstealing] def transitionWorkerFromSearching(random: ThreadLocalRandom): Unit = {
+    // Decrement the number of searching worker threads.
+    val prev = state.getAndDecrement()
+    if (prev == 1) {
+      // If this was the only searching thread, wake a thread up to potentially help out
+      // with the local work queue.
+      notifyParked(random)
+      ()
+    }
+  }
+
+  /** Updates the internal state to mark the given worker thread as parked.
+    *
+    * @return `true` if the given worker thread was the last searching thread,
+    *         `false` otherwise
+    */
+  private[workstealing] def transitionWorkerToParkedWhenSearching(): Boolean = {
+    // Decrement the number of unparked and searching threads simultaneously
+    // since this thread was searching prior to parking.
+    val prev = state.getAndAdd(-DeltaSearching)
+    (prev & SearchMask) == 1
+  }
+
+  /** Updates the internal state to mark the given worker thread as parked.
+    *
+    * @note This method is intentionally duplicated, to accomodate the
+    *       unconditional code paths in the [[WorkerThread]] runloop.
+    */
+  private[workstealing] def transitionWorkerToParked(): Unit = {
+    // Decrement the number of unparked threads only.
+    state.getAndAdd(-DeltaNotSearching)
+    ()
+  }
+
+  /** Executes a fiber on this thread pool.
+    *
+    * If the request comes from a [[WorkerThread]], the fiber is enqueued on the
+    * local queue of that thread.
+    *
+    * If the request comes from a [[HelperTread]] or an external thread, the
+    * fiber is enqueued on the overflow queue. Furthermore, if the request comes
+    * from an external thread, worker threads are notified of new work.
+    *
+    * @param fiber the fiber to be executed on the thread pool
+    */
+  private[this] def executeFiber(fiber: Runnable): Unit = {
+    val thread = Thread.currentThread()
+    if (thread.isInstanceOf[WorkerThread]) {
+      thread.asInstanceOf[WorkerThread].schedule(fiber)
+    } else if (thread.isInstanceOf[HelperThread]) {
+      thread.asInstanceOf[HelperThread].schedule(fiber)
+    } else {
+      val random = ThreadLocalRandom.current()
+      overflowQueue.offer(fiber, random)
+      notifyParked(random)
+      ()
+    }
+  }
+
+  /** Reschedules the given fiber directly on the local work stealing queue on the same thread,
+    * but with the possibility to skip notifying other fibers of a potential steal target, which
+    * reduces contention in workloads running on fewer worker threads. This method executes an
+    * unchecked cast to a `WorkerThread` and should only ever be called directly from a
+    * `WorkerThread`.
+    */
+  private[this] def rescheduleFiber(fiber: Runnable): Unit = {
```

Comment:
Why are these `private[this]`? Aren't `scheduleFiber` and `rescheduleFiber` supposed to be used?

Or maybe we'll expose them after modifying `Scheduler`.

Reply:
I really only did a single pass over the code,  this is far from ready. I guess it can be commented out for now. They definitely cannot be used at the moment. :/